lxdprofile.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+	lxdapi "github.com/lxc/lxd/shared/api"
+	"gopkg.in/yaml.v2"
+)
+
+type LXDProfile struct {
+	lxdapi.Profile
+}
+
+func NewLXDProfile() *LXDProfile {
+	return &LXDProfile{}
+}
+
+// ReadLXDProfile reads in a LXDProfile from a charm's lxd-profile.yaml.
+// It is not validated at this point so that the caller can choose to override
+// any validation.
+func ReadLXDProfile(r io.Reader) (*LXDProfile, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	var profile LXDProfile
+	if err := yaml.Unmarshal(data, &profile.Profile); err != nil {
+		return nil, errors.Annotate(err, "failed to unmarshall lxd-profile.yaml")
+	}
+	return &profile, nil
+}
+
+// HasName returns true if the lxd profile name is set
+func (profile *LXDProfile) HasName() bool {
+	return profile.Name != ""
+}
+
+// ValidateConfigDevices validates the Config and Devices properties of the LXDProfile.
+// WhiteList devices: unix-char, unix-block, gpu, usb.
+// BlackList config: boot*, limits* and migration*.
+func (profile *LXDProfile) ValidateConfigDevices() error {
+	if len(profile.Devices) < 1 && len(profile.Config) < 1 {
```

Comment:
It would be nice if we could ask if the profile is empty before validating the LXDProfile. The way the parsing of the file is done, the LXDProfile will always return a non-nil ptr (inline with other types inside charm code). So it would nice to either have `charm.IsLXDProfileEmpty()` or `lxdProfile.Empty()`, which can tell us if any items within the profile have been defined or not.

Reply:
adding lxdProfile.Empty() to keep in line with the other pieces of charm.